src/time-wait-sync/time-wait-sync.c

```diff
@@ -0,0 +1,176 @@
+        sd_event_source *event_source; /* non-null is the active io event source */
+} ClockState;
+
+static void clock_state_release(ClockState *sp) {
+        sp->event_source = sd_event_source_unref(sp->event_source);
+        sp->fd = safe_close(sp->fd);
+}
+
+static int clock_state_update(ClockState *sp,
+                              sd_event *event);
+
+static int io_handler(sd_event_source * s,
+                      int fd,
+                      uint32_t revents,
+                      void *userdata) {
+        ClockState *sp = userdata;
+
+        return clock_state_update(sp, sd_event_source_get_event(s));
+}
+
+static int clock_state_update(ClockState *sp,
+                              sd_event *event) {
+        int r;
+        static const struct itimerspec its = {
+                .it_value.tv_sec = TIME_T_MAX,
+        };
+        struct timex tx = {};
+        char buf[MAX(FORMAT_TIMESTAMP_MAX, 64)];
+        usec_t t;
+        const char * ts;
+
+        clock_state_release(sp);
+
+        r = timerfd_create(CLOCK_REALTIME, TFD_NONBLOCK | TFD_CLOEXEC);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create timerfd: %m");
+                goto finish;
+        }
+        sp->fd = r;
+
+        r = timerfd_settime(sp->fd, TFD_TIMER_ABSTIME | TFD_TIMER_CANCEL_ON_SET, &its, NULL);
+        if (r < 0) {
+                log_error_errno(r, "Failed to set timerfd conditions: %m");
+                goto finish;
+        }
+
+        r = adjtimex(&tx);
+        if (r < 0) {
+                log_error_errno(r, "Failed to read adjtimex state: %m");
+                goto finish;
+        }
+        sp->adjtime_state = r;
+
+        if (tx.status & STA_NANO) {
+                tx.time.tv_usec /= 1000;
+        }
+        t = timeval_load(&tx.time);
+        ts = format_timestamp_us_utc(buf, sizeof(buf), t);
+        if (!ts) {
+                strcpy(buf, "unrepresentable");
+        }
+        fprintf(stderr, "adjtime state %d status %x time %s\n", sp->adjtime_state, tx.status, ts);
+
+        if (sp->adjtime_state == TIME_ERROR) {
+                /* Not synchronized.  Do a one-shot wait on the descriptor and inform the caller we need to keep
+                 * running. */
+                r = sd_event_add_io(event, &sp->event_source, sp->fd,
+                                    EPOLLIN, io_handler, sp);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to create time change monitor source: %m");
+                        goto finish;
+                }
+                r = 1;
+        } else {
+                /* Synchronized; we can exit. */
+                (void) sd_event_exit(event, 0);
+                r = 0;
+        }
+
+ finish:
+        if (r < 0)
+                (void) sd_event_exit(event, r);
+        return r;
+}
+
+int main(int argc,
+         char * argv[]) {
+        int r;
+        sd_event *event;
```

Comment:
consider using `_cleanup_(sd_event_unrefp)` for this, so that you do not need to explicitl free this